File: packages/singular/ios/Plugin/SingularPlugin.swift
import Foundation
import Capacitor

@objc(SingularPlugin)
public class SingularPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SingularPlugin"
    public let jsName = "Singular"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "clearGlobalProperties", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createReferrerShortLink", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getGlobalProperties", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getLimitDataSharing", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAllTrackingStopped", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resumeAllTracking", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setCustomUserId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setDeviceToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setGlobalProperty", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setLimitAdvertisingIdentifiers", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setLimitDataSharing", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "skanGetConversionValue", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "skanRegisterAppForAdNetworkAttribution", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "skanUpdateConversionValue", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopAllTracking", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackAdRevenue", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackRevenue", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackingOptIn", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackingUnder13", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unsetCustomUserId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unsetGlobalProperty", returnType: CAPPluginReturnPromise)
    ]
    public static let eventDeviceAttributionInfoReceived = "deviceAttributionInfoReceived"
    public static let eventSingularLinkResolved = "singularLinkResolved"
    public static let eventSkanConversionValueUpdated = "skanConversionValueUpdated"
    public let tag = "Singular"
    private var implementation: SingularImpl?

    override public func load() {
        super.load()
        self.implementation = SingularImpl(plugin: self)
    }

    @objc func clearGlobalProperties(_ call: CAPPluginCall) {
        do {
            try implementation?.clearGlobalProperties(completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func createReferrerShortLink(_ call: CAPPluginCall) {
        do {
            let options = try CreateReferrerShortLinkOptions(call)
            try implementation?.createReferrerShortLink(options, completion: { result, error in
                self.handleCompletion(call, result, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func getGlobalProperties(_ call: CAPPluginCall) {
        do {
            try implementation?.getGlobalProperties(completion: { result, error in
                self.handleCompletion(call, result, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func getLimitDataSharing(_ call: CAPPluginCall) {
        do {
            try implementation?.getLimitDataSharing(completion: { result, error in
                self.handleCompletion(call, result, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func initialize(_ call: CAPPluginCall) {
        do {
            let options = try InitializeOptions(call)
            implementation?.initialize(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func isAllTrackingStopped(_ call: CAPPluginCall) {
        do {
            try implementation?.isAllTrackingStopped(completion: { result, error in
                self.handleCompletion(call, result, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    public func notifyDeviceAttributionInfoReceivedListeners(_ event: DeviceAttributionInfoReceivedEvent) {
        notifyListeners(Self.eventDeviceAttributionInfoReceived, data: event.toJSObject() as? [String: Any])
    }

    public func notifySingularLinkResolvedListeners(_ event: SingularLinkResolvedEvent) {
        notifyListeners(Self.eventSingularLinkResolved, data: event.toJSObject() as? [String: Any])
    }

    public func notifySkanConversionValueUpdatedListeners(_ event: SkanConversionValueUpdatedEvent) {
        notifyListeners(Self.eventSkanConversionValueUpdated, data: event.toJSObject() as? [String: Any])
    }

    @objc func resumeAllTracking(_ call: CAPPluginCall) {
        do {
            try implementation?.resumeAllTracking(completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func setCustomUserId(_ call: CAPPluginCall) {
        do {
            let options = try SetCustomUserIdOptions(call)
            try implementation?.setCustomUserId(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func setDeviceToken(_ call: CAPPluginCall) {
        do {
            let options = try SetDeviceTokenOptions(call)
            try implementation?.setDeviceToken(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func setGlobalProperty(_ call: CAPPluginCall) {
        do {
            let options = try SetGlobalPropertyOptions(call)
            try implementation?.setGlobalProperty(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func setLimitAdvertisingIdentifiers(_ call: CAPPluginCall) {
        do {
            let options = try SetLimitAdvertisingIdentifiersOptions(call)
            try implementation?.setLimitAdvertisingIdentifiers(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func setLimitDataSharing(_ call: CAPPluginCall) {
        do {
            let options = try SetLimitDataSharingOptions(call)
            try implementation?.setLimitDataSharing(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func skanGetConversionValue(_ call: CAPPluginCall) {
        do {
            try implementation?.skanGetConversionValue(completion: { result, error in
                self.handleCompletion(call, result, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func skanRegisterAppForAdNetworkAttribution(_ call: CAPPluginCall) {
        do {
            try implementation?.skanRegisterAppForAdNetworkAttribution(completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func skanUpdateConversionValue(_ call: CAPPluginCall) {
        do {
            let options = try SkanUpdateConversionValueOptions(call)
            try implementation?.skanUpdateConversionValue(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func stopAllTracking(_ call: CAPPluginCall) {
        do {
            try implementation?.stopAllTracking(completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func trackAdRevenue(_ call: CAPPluginCall) {
        do {
            let options = try TrackAdRevenueOptions(call)
            try implementation?.trackAdRevenue(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func trackEvent(_ call: CAPPluginCall) {
        do {
            let options = try TrackEventOptions(call)
            try implementation?.trackEvent(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func trackRevenue(_ call: CAPPluginCall) {
        do {
            let options = try TrackRevenueOptions(call)
            try implementation?.trackRevenue(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func trackingOptIn(_ call: CAPPluginCall) {
        do {
            try implementation?.trackingOptIn(completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func trackingUnder13(_ call: CAPPluginCall) {
        do {
            try implementation?.trackingUnder13(completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func unsetCustomUserId(_ call: CAPPluginCall) {
        do {
            try implementation?.unsetCustomUserId(completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func unsetGlobalProperty(_ call: CAPPluginCall) {
        do {
            let options = try UnsetGlobalPropertyOptions(call)
            try implementation?.unsetGlobalProperty(options, completion: { error in
                self.handleCompletion(call, error)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    private func handleCompletion(_ call: CAPPluginCall, _ error: Error?) {
        if let error = error {
            rejectCall(call, error)
        } else {
            resolveCall(call)
        }
    }

    private func handleCompletion(_ call: CAPPluginCall, _ result: Result?, _ error: Error?) {
        if let error = error {
            rejectCall(call, error)
        } else {
            resolveCall(call, result)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", self.tag, "] ", error)
        call.reject(error.localizedDescription, (error as? CustomError)?.code)
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}

File: packages/singular/android/src/main/java/io/capawesome/capacitorjs/plugins/singular/SingularPlugin.java
package io.capawesome.capacitorjs.plugins.singular;

import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.singular.classes.CustomException;
import io.capawesome.capacitorjs.plugins.singular.classes.events.DeviceAttributionInfoReceivedEvent;
import io.capawesome.capacitorjs.plugins.singular.classes.events.SingularLinkResolvedEvent;
import io.capawesome.capacitorjs.plugins.singular.classes.options.CreateReferrerShortLinkOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.InitializeOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.SetCustomUserIdOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.SetDeviceTokenOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.SetGlobalPropertyOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.SetLimitAdvertisingIdentifiersOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.SetLimitDataSharingOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.TrackAdRevenueOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.TrackEventOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.TrackRevenueOptions;
import io.capawesome.capacitorjs.plugins.singular.classes.options.UnsetGlobalPropertyOptions;
import io.capawesome.capacitorjs.plugins.singular.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.singular.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.singular.interfaces.Result;

@CapacitorPlugin(name = "Singular")
public class SingularPlugin extends Plugin {

    public static final String EVENT_DEVICE_ATTRIBUTION_INFO_RECEIVED = "deviceAttributionInfoReceived";
    public static final String EVENT_SINGULAR_LINK_RESOLVED = "singularLinkResolved";
    public static final String TAG = "SingularPlugin";

    private static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";

    private Singular implementation;

    @Override
    public void load() {
        super.load();
        this.implementation = new Singular(this);
    }

    @PluginMethod
    public void clearGlobalProperties(PluginCall call) {
        try {
            implementation.clearGlobalProperties(createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void createReferrerShortLink(PluginCall call) {
        try {
            CreateReferrerShortLinkOptions options = new CreateReferrerShortLinkOptions(call);
            implementation.createReferrerShortLink(options, createNonEmptyResultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getGlobalProperties(PluginCall call) {
        try {
            implementation.getGlobalProperties(createNonEmptyResultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getLimitDataSharing(PluginCall call) {
        try {
            implementation.getLimitDataSharing(createNonEmptyResultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        try {
            InitializeOptions options = new InitializeOptions(call);
            implementation.initialize(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void isAllTrackingStopped(PluginCall call) {
        try {
            implementation.isAllTrackingStopped(createNonEmptyResultCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    public void notifyDeviceAttributionInfoReceivedListeners(@NonNull DeviceAttributionInfoReceivedEvent event) {
        notifyListeners(EVENT_DEVICE_ATTRIBUTION_INFO_RECEIVED, event.toJSObject());
    }

    public void notifySingularLinkResolvedListeners(@NonNull SingularLinkResolvedEvent event) {
        notifyListeners(EVENT_SINGULAR_LINK_RESOLVED, event.toJSObject());
    }

    @PluginMethod
    public void resumeAllTracking(PluginCall call) {
        try {
            implementation.resumeAllTracking(createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void setCustomUserId(PluginCall call) {
        try {
            SetCustomUserIdOptions options = new SetCustomUserIdOptions(call);
            implementation.setCustomUserId(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void setDeviceToken(PluginCall call) {
        try {
            SetDeviceTokenOptions options = new SetDeviceTokenOptions(call);
            implementation.setDeviceToken(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void setGlobalProperty(PluginCall call) {
        try {
            SetGlobalPropertyOptions options = new SetGlobalPropertyOptions(call);
            implementation.setGlobalProperty(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void setLimitAdvertisingIdentifiers(PluginCall call) {
        try {
            SetLimitAdvertisingIdentifiersOptions options = new SetLimitAdvertisingIdentifiersOptions(call);
            implementation.setLimitAdvertisingIdentifiers(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void setLimitDataSharing(PluginCall call) {
        try {
            SetLimitDataSharingOptions options = new SetLimitDataSharingOptions(call);
            implementation.setLimitDataSharing(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void skanGetConversionValue(PluginCall call) {
        rejectCallAsUnimplemented(call);
    }

    @PluginMethod
    public void skanRegisterAppForAdNetworkAttribution(PluginCall call) {
        rejectCallAsUnimplemented(call);
    }

    @PluginMethod
    public void skanUpdateConversionValue(PluginCall call) {
        rejectCallAsUnimplemented(call);
    }

    @PluginMethod
    public void stopAllTracking(PluginCall call) {
        try {
            implementation.stopAllTracking(createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void trackAdRevenue(PluginCall call) {
        try {
            TrackAdRevenueOptions options = new TrackAdRevenueOptions(call);
            implementation.trackAdRevenue(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void trackEvent(PluginCall call) {
        try {
            TrackEventOptions options = new TrackEventOptions(call);
            implementation.trackEvent(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void trackRevenue(PluginCall call) {
        try {
            TrackRevenueOptions options = new TrackRevenueOptions(call);
            implementation.trackRevenue(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void trackingOptIn(PluginCall call) {
        try {
            implementation.trackingOptIn(createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void trackingUnder13(PluginCall call) {
        try {
            implementation.trackingUnder13(createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void unsetCustomUserId(PluginCall call) {
        try {
            implementation.unsetCustomUserId(createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void unsetGlobalProperty(PluginCall call) {
        try {
            UnsetGlobalPropertyOptions options = new UnsetGlobalPropertyOptions(call);
            implementation.unsetGlobalProperty(options, createEmptyCallback(call));
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @Override
    protected void handleOnNewIntent(Intent intent) {
        super.handleOnNewIntent(intent);
        implementation.handleNewIntent(intent);
    }

    @NonNull
    private EmptyCallback createEmptyCallback(@NonNull PluginCall call) {
        return new EmptyCallback() {
            @Override
            public void success() {
                resolveCall(call);
            }

            @Override
            public void error(Exception exception) {
                rejectCall(call, exception);
            }
        };
    }

    @NonNull
    private <T extends Result> NonEmptyResultCallback<T> createNonEmptyResultCallback(@NonNull PluginCall call) {
        return new NonEmptyResultCallback<T>() {
            @Override
            public void success(@NonNull T result) {
                resolveCall(call, result);
            }

            @Override
            public void error(Exception exception) {
                rejectCall(call, exception);
            }
        };
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void rejectCallAsUnimplemented(@NonNull PluginCall call) {
        call.unimplemented("This method is not available on this platform.");
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}
